gcc/d/d-decls.cc

```diff
@@ -28,753 +28,898 @@
+  InterfaceDeclaration *id = decl->isInterfaceDeclaration ();
+  tree ident = make_internal_name (decl, id ? "__Interface" : "__Class", "Z");
 
-      setup_symbol_storage (this, decl, true);
-      set_decl_location (decl, this);
+  decl->csym = build_decl (BUILTINS_LOCATION, VAR_DECL,
+			   IDENTIFIER_PRETTY_NAME (ident), unknown_type_node);
+  set_decl_location (decl->csym, decl);
+  DECL_LANG_SPECIFIC (decl->csym) = build_lang_decl (NULL);
+  SET_DECL_ASSEMBLER_NAME (decl->csym, ident);
 
-      DECL_ARTIFICIAL (decl) = 1;
-      // Not readonly, moduleinit depends on this.
-      TREE_READONLY (decl) = 0;
-    }
+  d_keep (decl->csym);
 
-  return csym;
-}
+  /* Class is a reference, want the record type.  */
+  DECL_CONTEXT (decl->csym) = TREE_TYPE (build_ctype (decl->type));
+  DECL_ARTIFICIAL (decl->csym) = 1;
+  TREE_STATIC (decl->csym) = 1;
+  /* ClassInfo cannot be const data, because we use the monitor on it.  */
+  TREE_READONLY (decl->csym) = 0;
+  TREE_PUBLIC (decl->csym) = 1;
 
-Symbol *
-StructLiteralExp::toSymbol()
-{
-  if (!sym)
-    {
-      sym = new Symbol();
+  /* The moduleinfo decl has not been defined -- yet.  */
+  DECL_EXTERNAL (decl->csym) = 1;
 
-      // Build reference symbol.
-      tree ctype = build_ctype(type);
-      tree decl = build_artificial_decl(ctype, NULL_TREE, "S");
-      set_decl_location(decl, loc);
+  /* Could move setting comdat linkage to the caller, who knows whether
+     this classinfo is being emitted in this compilation.  */
+  if (decl->isInstantiated ())
+    d_comdat_linkage (decl->csym);
 
-      sym->Stree = decl;
-      this->sinit = sym;
+  return decl->csym;
+}
 
-      DECL_INITIAL (sym->Stree) = build_expr(this, true);
-      d_finish_symbol(sym);
-    }
+/* Get the VAR_DECL of the vtable symbol for DECL.  If this does not yet exist,
+   create it.  The vtable is accessible via ClassInfo, but since it is needed
+   frequently (like for rtti comparisons), make it directly accessible.  */
 
-  return sym;
+tree
+get_vtable_decl (ClassDeclaration *decl)
+{
+  if (decl->vtblsym)
+    return decl->vtblsym;
+
+  /* Note: Using a static array type for the VAR_DECL, the DECL_INITIAL value
+     will have a different type.  However the backend seems to accept this.  */
+  Type *vtbltype = Type::tvoidptr->sarrayOf (decl->vtbl.dim);
+  tree ident = make_internal_name (decl, "__vtbl", "Z");
+
+  decl->vtblsym = build_decl (UNKNOWN_LOCATION, VAR_DECL,
+			      IDENTIFIER_PRETTY_NAME (ident),
+			      build_ctype (vtbltype));
+  set_decl_location (decl->vtblsym, decl);
+  DECL_LANG_SPECIFIC (decl->vtblsym) = build_lang_decl (NULL);
+  SET_DECL_ASSEMBLER_NAME (decl->vtblsym, ident);
+
+  d_keep (decl->vtblsym);
+
+  /* Class is a reference, want the record type.  */
+  DECL_CONTEXT (decl->vtblsym) = TREE_TYPE (build_ctype (decl->type));
+  DECL_ARTIFICIAL (decl->vtblsym) = 1;
+  TREE_STATIC (decl->vtblsym) = 1;
+  TREE_READONLY (decl->vtblsym) = 1;
+  DECL_VIRTUAL_P (decl->vtblsym) = 1;
```

Comment:
There's an annoying checking assert in GCC <= 4.9 related to this:
```c++
  /* Vtables are defined by their types and must match no matter of interposition
     rules.  */
  if (DECL_VIRTUAL_P (real_decl))
    {
      gcc_checking_assert (TREE_READONLY (real_decl));
      if (DECL_INITIAL (real_decl))
	return DECL_INITIAL (real_decl);
      else
	{
	  /* The C++ front end creates VAR_DECLs for vtables of typeinfo
	     classes not defined in the current TU so that it can refer
	     to them from typeinfo objects.  Avoid returning NULL_TREE.  */
	  gcc_checking_assert (!COMPLETE_TYPE_P (DECL_CONTEXT (real_decl)));
	  return error_mark_node;
	}
    }
```
I assume I can simply remove the `DECL_VIRTUAL_P (decl->vtblsym) = 1;` line?